src/control/server/instance.go

```diff
@@ -218,9 +218,47 @@ func (ei *EngineInstance) determineRank(ctx context.Context, ready *srvpb.Notify
 	return r, resp.LocalJoin, nil
 }
 
+func (ei *EngineInstance) isFaultDomainChanging(sb *Superblock) bool {
+	if ei.hostFaultDomain == nil && sb.HostFaultDomain == "" {
+		return false
+	}
+
+	return !(ei.hostFaultDomain.String() == sb.HostFaultDomain)
+}
+
+func (ei *EngineInstance) updateFaultDomainInSuperblock() error {
+	superblock := ei.getSuperblock()
+	if superblock == nil {
+		return errors.New("nil superblock while updating fault domain")
+	}
+
+	if !ei.isFaultDomainChanging(superblock) {
+		return nil
+	}
+
+	ei.log.Infof("instance %d setting host fault domain to %q (previously %q)",
+		ei.Index(), ei.hostFaultDomain, superblock.HostFaultDomain)
+
+	newDomainStr := ""
+	if ei.hostFaultDomain != nil {
+		newDomainStr = ei.hostFaultDomain.String()
+	}
+	superblock.HostFaultDomain = newDomainStr
```

Comment:
This seems a bit confusing... Why would we potentially allow an empty fault domain to be persisted to the superblock? I would think instead that we should return an error if `ei.hostFaultDomain` is nil or otherwise resolves to an empty string.

Reply:
Good point. I'll make that change.